### ISSUE TITLE
netbox-dns installation in a docker compose Environment fails

### ISSUE THREAD

**Issue by lairsdragon** [peteeckel/netbox-plugin-dns#138]
**Versions**
NetBox Version: 3.6.9
NetBox DNS Version: 0.21.14
Python Version: 3.11.4

**Describe the bug**
I am running netbox from `docker compose` environment and part of the that is container build to add the needed Plugins to the container. Last command in the Dockerfile ist a `/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py collectstatic --no-input` which fails because the `netbox_dns` plugin tries to open the Database and fail. Since this in a build phase there is no databate which could be connected at that point in time..

**To Reproduce**
1. Start of with https://github.com/netbox-community/netbox-docker/releases/tag/2.7.0
2. Add `netbox-plugin-dns` to `plugin_requirements.txt`
3. Run a `docker compose build` 
4. See Build Error
   `failed to solve: process "/bin/sh -c SECRET_KEY=\"Souch8Fis6mongei9yaezeeto4aini5Johleip5ee9ooxieC9v\" /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py collectstatic --no-input" did not complete successfully: exit code: 1`

**Expected result**
A call to `/opt/netbox/netbox/manage.py collectstatic` should not open the database so that this step can used in build process as it works without the dns plugin 

**Actual result**
The build process fails and it is not possible to build a docker container including the dns plugin 

**Screenshots**
```
root@steve-docker-host:/opt/docker-compose/netbox-docker# docker compose build
[+] Building 5.3s (9/9) FINISHED                                                                                                                                                                    docker:default
 => [netbox internal] load build definition from Dockerfile-Plugins                                                                                                                                           0.0s
 => => transferring dockerfile: 508B                                                                                                                                                                          0.0s
 => [netbox internal] load .dockerignore                                                                                                                                                                      0.0s
 => => transferring context: 162B                                                                                                                                                                             0.0s
 => [netbox internal] load metadata for docker.io/netboxcommunity/netbox:v3.6-2.7.0                                                                                                                           1.1s
 => [netbox 1/5] FROM docker.io/netboxcommunity/netbox:v3.6-2.7.0@sha256:cf8a16f61d96c5d15cf5c639129662390e4b42685ed9c94ab4e344edeceb9a70                                                                     0.0s
 => [netbox internal] load build context                                                                                                                                                                      0.0s
 => => transferring context: 128B                                                                                                                                                                             0.0s
 => CACHED [netbox 2/5] COPY ./plugin_requirements.txt /                                                                                                                                                      0.0s
 => CACHED [netbox 3/5] RUN /opt/netbox/venv/bin/pip install  --no-warn-script-location -r /plugin_requirements.txt                                                                                           0.0s
 => CACHED [netbox 4/5] COPY configuration/configuration.py /etc/netbox/config/configuration.py                                                                                                               0.0s
 => ERROR [netbox 5/5] RUN SECRET_KEY="xxx" /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py collectstatic --no-input                                  4.1s
------                                                                                                                                                                                                             
 > [netbox 5/5] RUN SECRET_KEY="xxx" /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py collectstatic --no-input:                                             
1.384 🧬 loaded config '/etc/netbox/config/configuration.py'                                                                                                                                                       
1.385 🧬 loaded config '/etc/netbox/config/extra.py'                                                                                                                                                               
1.385 🧬 loaded config '/etc/netbox/config/logging.py'                                                                                                                                                             
1.387 🧬 loaded config '/etc/netbox/config/plugins.py'                                                                                                                                                             
3.616 Traceback (most recent call last):
3.616   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
3.618     self.connect()
3.618   File "/opt/netbox/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
3.618     return func(*args, **kwargs)
3.618            ^^^^^^^^^^^^^^^^^^^^^
3.619   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 270, in connect
3.621     self.connection = self.get_new_connection(conn_params)
3.621                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.625   File "/opt/netbox/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
3.625     return func(*args, **kwargs)
3.626            ^^^^^^^^^^^^^^^^^^^^^
3.632   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/postgresql/base.py", line 275, in get_new_connection
3.632     connection = self.Database.connect(**conn_params)
3.632                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.632   File "/opt/netbox/venv/lib/python3.11/site-packages/psycopg/connection.py", line 729, in connect
3.632     raise ex.with_traceback(None)
3.633 psycopg.OperationalError: connection failed: Connection refused
3.635 	Is the server running on that host and accepting TCP/IP connections?
3.635 connection to server at "localhost" (::1), port 5432 failed: Cannot assign requested address
3.635 	Is the server running on that host and accepting TCP/IP connections?
3.635 
3.635 The above exception was the direct cause of the following exception:
3.635 
3.635 Traceback (most recent call last):
3.635   File "/opt/netbox/netbox/manage.py", line 10, in <module>
3.635     execute_from_command_line(sys.argv)
3.635   File "/opt/netbox/venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
3.635     utility.execute()
3.635   File "/opt/netbox/venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 416, in execute
3.635     django.setup()
3.635   File "/opt/netbox/venv/lib/python3.11/site-packages/django/__init__.py", line 24, in setup
3.635     apps.populate(settings.INSTALLED_APPS)
3.635   File "/opt/netbox/venv/lib/python3.11/site-packages/django/apps/registry.py", line 124, in populate
3.636     app_config.ready()
3.636   File "/opt/netbox/venv/lib/python3.11/site-packages/netbox_dns/__init__.py", line 53, in ready
3.636     objtype = ContentType.objects.get_for_model(IPAddress)
3.636               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/contrib/contenttypes/models.py", line 52, in get_for_model
3.644     ct = self.get(app_label=opts.app_label, model=opts.model_name)
3.644          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
3.644     return getattr(self.get_queryset(), name)(*args, **kwargs)
3.644            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/query.py", line 633, in get
3.644     num = len(clone)
3.644           ^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/query.py", line 380, in __len__
3.644     self._fetch_all()
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/query.py", line 1881, in _fetch_all
3.644     self._result_cache = list(self._iterable_class(self))
3.644                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/query.py", line 91, in __iter__
3.644     results = compiler.execute_sql(
3.644               ^^^^^^^^^^^^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/sql/compiler.py", line 1560, in execute_sql
3.644     cursor = self.connection.cursor()
3.644              ^^^^^^^^^^^^^^^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
3.644     return func(*args, **kwargs)
3.644            ^^^^^^^^^^^^^^^^^^^^^
3.644   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 330, in cursor
3.644     return self._cursor()
3.644            ^^^^^^^^^^^^^^
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 306, in _cursor
3.650     self.ensure_connection()
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
3.650     return func(*args, **kwargs)
3.650            ^^^^^^^^^^^^^^^^^^^^^
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 288, in ensure_connection
3.650     with self.wrap_database_errors:
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/utils.py", line 91, in __exit__
3.650     raise dj_exc_value.with_traceback(traceback) from exc_value
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
3.650     self.connect()
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
3.650     return func(*args, **kwargs)
3.650            ^^^^^^^^^^^^^^^^^^^^^
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/base/base.py", line 270, in connect
3.650     self.connection = self.get_new_connection(conn_params)
3.650                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.650   File "/opt/netbox/venv/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
3.650     return func(*args, **kwargs)
3.652            ^^^^^^^^^^^^^^^^^^^^^
3.652   File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/backends/postgresql/base.py", line 275, in get_new_connection
3.652     connection = self.Database.connect(**conn_params)
3.652                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
3.652   File "/opt/netbox/venv/lib/python3.11/site-packages/psycopg/connection.py", line 729, in connect
3.652     raise ex.with_traceback(None)
3.652 django.db.utils.OperationalError: connection failed: Connection refused
3.652 	Is the server running on that host and accepting TCP/IP connections?
3.652 connection to server at "localhost" (::1), port 5432 failed: Cannot assign requested address
3.652 	Is the server running on that host and accepting TCP/IP connections?
------
failed to solve: process "/bin/sh -c SECRET_KEY=\"Souch8Fis6mongei9yaezeeto4aini5Johleip5ee9ooxieC9v\" /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py collectstatic --no-input" did not complete successfully: exit code: 1
root@steve-docker-host:/opt/docker-compose/netbox-docker# cat Dockerfile-Plugins 
FROM netboxcommunity/netbox:v3.6-2.7.0

COPY ./plugin_requirements.txt /
RUN /opt/netbox/venv/bin/pip install  --no-warn-script-location -r /plugin_requirements.txt

# These lines are only required if your plugin has its own static files.
COPY configuration/configuration.py /etc/netbox/config/configuration.py
RUN SECRET_KEY="xxx" /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py collectstatic --no-input

```

**Comment by peteeckel** [comment]
Hi @lairsdragon, thanks for raising this.

First things first, I don't use Docker at all, so this might be a bit hard for me to reproduce.

NetBox DNS does not bring any static files at all, so I don't really see why the presence of the plugin should affect whether `manage.py collectstatic` requires a database connection or not. There is also no way in which I could influence this from the plugin code.

It seems to be a Django-specific issue according to [this](https://stackoverflow.com/questions/74121134/django-collectstatic-without-db-connection) conversation on Stackoverflow. There's also a workaround (untested).

**Comment by peteeckel** [comment]
I investigated a bit more deeply, and it seems it is in fact an issue with how `manage.py` is started.

NetBox DNS runs the following command on initialisation:
```
objtype = ContentType.objects.get_for_model(IPAddress)
```

That obviously requires a database connection, and do it fails in your setup when the database is not running.

The code is actually in the `ready()` method, which determines whether all required dynamic fields for IPAM coupling are present (and contains an unrelated bug, by the way). This check isn't strictly necessary, so one way to get rid of the problem would be to remove it altogether.

However, I can't say at this moment whether there is something else elsewhere that still causes the problem to occur.

So the workaround from the Stackoverflow conversation might be the safer way to solve it (provided that it works).

**Comment by lairsdragon** [comment]
Thank for looking in that issue,

It only happens if feature_ipam_coupling was set. (Which I have done, after the first sucessful container build)

I suppose you mean the SO Answer with SKIP_PG_EXTRA_VALIDATION? I think that would only work on the specific django plugin...

Maybe a way around would be to see if a specific Envirnoment Variable (eg: "NETBOX_DNS_SKIP_CHECKS") was set and skip the check in that case..

**Comment by peteeckel** [comment]
I'll have a look later this evening ... there will be some way to solve it, I'm sure.

Probably I can already reproduce it by stopping the database in a non-Docker installation. If that works, there should be a quick solution for you.

**Comment by peteeckel** [comment]
I found a simple and relatively elegant way to work around that issue. 

As I said, the test for the custom fields is not strictly necessary, it just logs a warning if some or all custom fields are missing. If the database can't be connected to that check fails.

The workaround I implemented now catches the exception caused by that and logs a different warning instead. I think that will be sufficient as well, and now the `ready()` method no longer fails and the `collectstatic` operation can continue.
